lib/Dialect/Calyx/Export/CalyxEmitter.cpp

```diff
@@ -699,32 +699,33 @@ void Emitter::emitPrimitiveExtern(hw::HWModuleExternOp op) {
 void Emitter::emitPrimitivePorts(hw::HWModuleExternOp op) {
   auto emitPorts = [&](auto ports, bool isInput) {
     os << LParen();
-    for (size_t i = 0, e = ports.size(); i < e; ++i) {
-      const hw::PortInfo &port = ports[i];
+    for (auto [i, port] : llvm::enumerate(ports)) {
       DictionaryAttr portAttr =
           isInput ? op.getArgAttrDict(i) : op.getResultAttrDict(i);
 
       os << getAttributes(op, portAttr) << port.name.getValue() << colon();
       // We only care about the bit width in the emitted .futil file.
       // Emit parameterized or non-parameterized bit width.
       if (hw::isParametricType(port.type)) {
-        hw::ParamDeclRefAttr bitWidth = port.type.cast<hw::IntType>()
-                                            .getWidth()
-                                            .dyn_cast<hw::ParamDeclRefAttr>();
+        hw::ParamDeclRefAttr bitWidth =
+            port.type.template cast<hw::IntType>()
+                .getWidth()
+                .template dyn_cast<hw::ParamDeclRefAttr>();
         os << bitWidth.getName().str();
       } else {
         unsigned int bitWidth = port.type.getIntOrFloatBitWidth();
         os << bitWidth;
       }
 
-      if (i + 1 < e)
+      if (i + 1 < std::distance(ports.begin(), ports.end()))
```

Comment:
This seems less efficient. Perhaps calculate `e` outside of the loop?

Reply:
It's just a subtract, and probably is hoisted, but I moved it.

Reply:
Wouldn't surprise me if the compiler wasn't able to do either optimization and defaulted to iterating through with a counter (which would be necessary if it were a linked list). I don't think smallvector.begin() returns a random_access_iterator...

but whatever.

Reply:
no, vector's model random_access_iterator whose specialized distance is arithmatic.  It's not accidental or guessing that it's a subtration.